File: rust/src/irc/tracker.rs
use std::collections::BTreeSet;

use crate::irc::ISupportTracker;
use crate::prelude::*;

#[derive(Debug)]
pub struct Tracker {
    isupport: ISupportTracker,
    current_nick: Option<String>,
    channels: BTreeMap<String, ChannelState>,
}

#[derive(Debug, Clone)]
pub struct ChannelState {
    pub name: String,
    pub topic: Option<String>,
    pub users: BTreeSet<String>,
}

impl ChannelState {
    fn new(name: String) -> Self {
        ChannelState {
            name,
            topic: None,
            users: BTreeSet::new(),
        }
    }
}

impl Tracker {
    pub fn new() -> Self {
        Tracker {
            isupport: ISupportTracker::new(),
            current_nick: None,
            channels: BTreeMap::new(),
        }
    }

    pub fn get_current_nick(&self) -> Option<String> {
        self.current_nick.clone()
    }

    pub fn get_channel(&self, name: &str) -> Option<ChannelState> {
        self.channels.get(name).map(|val| val.clone())
    }

    pub fn list_channels(&self) -> Vec<String> {
        self.channels.keys().map(String::clone).collect()
    }

    pub fn handle_message(&mut self, msg: &irc::Message) {
        self.isupport.handle_message(msg);

        match (msg.command.as_str(), msg.params.len(), &msg.prefix) {
            ("001", 2, _) => self.handle_rpl_welcome(msg.params[0].as_str()),
            ("331", 3, _) => self.handle_rpl_no_topic(msg.params[1].as_str()),
            ("332", 3, _) => self.handle_rpl_topic(msg.params[1].as_str(), msg.params[2].as_str()),
            ("353", 4, _) => self
                .handle_rpl_nam_reply(msg.params[2].as_str(), msg.params[3].split(" ").collect()),
            ("JOIN", 1, Some(prefix)) => {
                self.handle_join(prefix.nick.as_str(), msg.params[0].as_str())
            }
            ("KICK", 3, _) => self.handle_kick(msg.params[1].as_str(), msg.params[0].as_str()),
            ("NICK", 1, Some(prefix)) => {
                self.handle_nick(prefix.nick.as_str(), msg.params[0].as_str())
            }
            ("PART", 2, Some(prefix)) => {
                self.handle_part(prefix.nick.as_str(), msg.params[0].as_str())
            }
            ("QUIT", 1, Some(prefix)) => self.handle_quit(prefix.nick.as_str()),
            ("TOPIC", 2, _) => self.handle_topic(msg.params[0].as_str(), msg.params[1].as_str()),

            _ => (),
        }
    }
}

impl Tracker {
    fn handle_rpl_welcome(&mut self, nick: &str) {
        self.current_nick = Some(nick.to_string());
    }

    fn handle_rpl_no_topic(&mut self, channel_name: &str) {
        if let Some(channel) = self.channels.get_mut(channel_name) {
            channel.topic = None;
        }
    }

    fn handle_rpl_topic(&mut self, channel_name: &str, topic: &str) {
        if let Some(channel) = self.channels.get_mut(channel_name) {
            channel.topic = Some(topic.to_string());
        }
    }

    fn handle_rpl_nam_reply(&mut self, channel_name: &str, users: Vec<&str>) {
        let prefix_map = self
            .isupport
            .get_prefix_map()
            .unwrap_or_else(|| BTreeMap::new());

        // The bot user should be added via JOIN so we need to filter it out. We
        // also use this chance to remove the unwanted prefixes.
        let users: Vec<_> = users
            .into_iter()
            .filter_map(|user| {
                if user.is_empty() || Some(user) == self.current_nick.as_deref() {
                    return None;
                }

                if prefix_map.contains_key(&user.as_bytes()[0]) {
                    Some(user[1..].to_string())
                } else {
                    Some(user.to_string())
                }
            })
            .collect();

        if let Some(channel) = self.channels.get_mut(channel_name) {
            for user in users.into_iter() {
                channel.users.insert(user.to_string());
            }
        } else {
            // Got RPL_NAMREPLY message for untracked channel
        }
    }

    fn handle_join(&mut self, nick: &str, channel_name: &str) {
        if !self.channels.contains_key(channel_name) && Some(nick) != self.current_nick.as_deref() {
            // Got JOIN message for untracked channel
            return;
        }

        let channel = self
            .channels
            .entry(channel_name.to_string())
            .or_insert_with(|| ChannelState::new(channel_name.to_string()));

        channel.users.insert(nick.to_string());
    }

    fn handle_kick(&mut self, nick: &str, channel_name: &str) {
        if !self.channels.contains_key(channel_name) {
            // Got KICK message for untracked channel
            return;
        }

        if Some(nick) == self.current_nick.as_deref() {
            self.channels.remove(channel_name);
        } else if let Some(channel) = self.channels.get_mut(channel_name) {
            channel.users.remove(nick);
        } else {
            unreachable!();
        }
    }

    fn handle_nick(&mut self, old_nick: &str, new_nick: &str) {
        for (_, channel) in self.channels.iter_mut() {
            if channel.users.remove(old_nick) {
                channel.users.insert(new_nick.to_string());
            }
        }
    }

    fn handle_part(&mut self, nick: &str, channel_name: &str) {
        if !self.channels.contains_key(channel_name) {
            // Got PART message for untracked channel
            return;
        }

        if Some(nick) == self.current_nick.as_deref() {
            self.channels.remove(channel_name);
        } else if let Some(channel) = self.channels.get_mut(channel_name) {
            channel.users.remove(nick);
        } else {
            unreachable!();
        }
    }

    fn handle_quit(&mut self, nick: &str) {
        for (_, channel) in self.channels.iter_mut() {
            channel.users.remove(nick);
        }
    }

    fn handle_topic(&mut self, channel_name: &str, topic: &str) {
        if let Some(channel) = self.channels.get_mut(channel_name) {
            channel.topic = Some(topic.to_string());
        }
    }
}

File: ircx/tracker.go
package ircx

// TODO: store all nicks by uuid and map them in outgoing seabird events rather
// than passing the nicks around directly

import (
	"strings"
	"sync"

	"github.com/sirupsen/logrus"
	"gopkg.in/irc.v3"
)

type Tracker struct {
	sync.RWMutex

	isupport *ISupportTracker
	channels map[string]*ChannelState
}

func NewTracker() *Tracker {
	return &Tracker{
		isupport: NewISupportTracker(),
		channels: make(map[string]*ChannelState),
	}
}

type ChannelState struct {
	Name  string
	Topic string
	Users map[string]struct{}
}

func (t *Tracker) ListChannels() []string {
	t.RLock()
	defer t.RUnlock()

	var ret []string
	for channel := range t.channels {
		ret = append(ret, channel)
	}

	return ret
}

func (t *Tracker) GetChannel(name string) *ChannelState {
	t.RLock()
	defer t.RUnlock()

	return t.channels[name]
}
func (t *Tracker) Handle(client *irc.Client, msg *irc.Message) {
	t.isupport.Handle(client, msg)

	switch msg.Command {
	case "332":
		t.handleRplTopic(client, msg)
	case "353":
		t.handleRplNamReply(client, msg)
	case "JOIN":
		t.handleJoin(client, msg)
	case "TOPIC":
		t.handleTopic(client, msg)
	case "PART":
		t.handlePart(client, msg)
	case "KICK":
		t.handleKick(client, msg)
	case "QUIT":
		t.handleQuit(client, msg)
	case "NICK":
		t.handleNick(client, msg)
	}
}

func (t *Tracker) handleTopic(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 2 {
		return
	}

	channel := msg.Params[0]
	topic := msg.Trailing()

	t.Lock()
	defer t.Unlock()

	if _, ok := t.channels[channel]; !ok {
		// Warning: got topic for channel we don't know about
		return
	}

	t.channels[channel].Topic = topic
}

func (t *Tracker) handleRplTopic(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 3 {
		return
	}

	// client := msg.Params[0]
	channel := msg.Params[1]
	topic := msg.Trailing()

	t.Lock()
	defer t.Unlock()

	if _, ok := t.channels[channel]; !ok {
		logrus.Warn("Got TOPIC message for untracked channel")
		return
	}

	logrus.WithFields(logrus.Fields{
		"channel": channel,
		"topic":   topic,
	}).Debug("Topic set in channel")

	t.channels[channel].Topic = topic
}

func (t *Tracker) handleJoin(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 1 {
		return
	}

	user := msg.Prefix.Name
	channel := msg.Trailing()

	t.Lock()
	defer t.Unlock()

	_, ok := t.channels[channel]

	if !ok {
		if user != client.CurrentNick() {
			logrus.Warn("Got JOIN message for untracked channel")
			return
		}

		logrus.WithFields(logrus.Fields{
			"channel": channel,
		}).Debug("Tracking channel")

		t.channels[channel] = &ChannelState{Name: channel, Users: make(map[string]struct{})}
	}

	logrus.WithFields(logrus.Fields{
		"nick":    user,
		"channel": channel,
	}).Debug("User joined channel")

	state := t.channels[channel]
	state.Users[user] = struct{}{}
}

func (t *Tracker) handlePart(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) < 1 {
		return
	}

	user := msg.Prefix.Name
	channel := msg.Params[0]

	t.Lock()
	defer t.Unlock()

	if _, ok := t.channels[channel]; !ok {
		logrus.Warn("Got PART message for untracked channel")
		return
	}

	if user == client.CurrentNick() {
		logrus.WithFields(logrus.Fields{
			"channel": channel,
		}).Debug("Done tracking channel")

		delete(t.channels, channel)
	} else {
		logrus.WithFields(logrus.Fields{
			"nick":    user,
			"channel": channel,
		}).Debug("User left channel")

		state := t.channels[channel]
		delete(state.Users, user)
	}
}

func (t *Tracker) handleKick(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 3 {
		return
	}

	actor := msg.Prefix.Name
	user := msg.Params[1]
	channel := msg.Params[0]

	t.Lock()
	defer t.Unlock()

	if _, ok := t.channels[channel]; !ok {
		logrus.Warn("Got KICK message for untracked channel")
		return
	}

	logrus.WithFields(logrus.Fields{
		"actor":   actor,
		"nick":    user,
		"channel": channel,
	}).Debug("User was kicked from channel")

	if user == client.CurrentNick() {
		delete(t.channels, channel)
	} else {
		state := t.channels[channel]
		delete(state.Users, user)
	}
}

func (t *Tracker) handleQuit(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 1 {
		return
	}

	user := msg.Prefix.Name

	logrus.WithFields(logrus.Fields{
		"nick": user,
	}).Debug("User has quit")

	t.Lock()
	defer t.Unlock()

	for _, state := range t.channels {
		delete(state.Users, user)
	}
}

func (t *Tracker) handleNick(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 1 {
		return
	}

	oldUser := msg.Prefix.Name
	newUser := msg.Params[0]

	logrus.WithFields(logrus.Fields{
		"old_nick": oldUser,
		"new_nick": newUser,
	}).Debug("Renaming user")

	t.Lock()
	defer t.Unlock()

	for _, state := range t.channels {
		if _, ok := state.Users[oldUser]; ok {
			delete(state.Users, oldUser)
			state.Users[newUser] = struct{}{}
		}
	}
}

func (t *Tracker) handleRplNamReply(client *irc.Client, msg *irc.Message) {
	if len(msg.Params) != 4 {
		return
	}

	channel := msg.Params[2]
	users := strings.Split(strings.TrimSpace(msg.Trailing()), " ")

	prefixes, ok := t.isupport.GetPrefixMap()
	if !ok {
		return
	}

	t.Lock()
	defer t.Unlock()

	if _, ok := t.channels[channel]; !ok {
		logrus.Warn("Got RPL_NAMREPLY message for untracked channel")
		return
	}

	for _, user := range users {
		i := strings.IndexFunc(user, func(r rune) bool {
			_, ok := prefixes[r]
			return !ok
		})

		if i != -1 {
			user = user[i:]
		}

		// The bot user should be added via JOIN
		if user == client.CurrentNick() {
			continue
		}

		state := t.channels[channel]
		state.Users[user] = struct{}{}
	}
}
